client/app/react-components/QueryEditor.js

```diff
@@ -0,0 +1,229 @@
+        switch (command) {
+          case 'focus': {
+            editor.focus();
+            break;
+          }
+          case 'paste': {
+            const [text] = args;
+            editor.session.doc.replace(editor.selection.getRange(), text);
+            const range = editor.selection.getRange();
+            this.props.updateQuery(editor.session.getValue());
+            window.setTimeout(() => {
+              editor.selection.setRange(range);
+            });
+            break;
+          }
+          default:
+            break;
+        }
+      });
+    };
+
+    this.formatQuery = () => {
+      // eslint-disable-next-line react/prop-types
+      const format = this.props.Query.format;
+      format(this.props.dataSource.syntax || 'sql', this.props.queryText)
+        .then(this.props.updateQuery)
+        .catch(error => toastr.error(error));
+    };
+  }
+
+  static getDerivedStateFromProps(nextProps, prevState) {
+    if (!nextProps.schema) {
+      return { keywords: [], autocompleteQuery: false };
+    } else if (nextProps.schema !== prevState.schema) {
+      return {
+        schema: nextProps.schema,
+        keywords: buildKeywordsFromSchema(nextProps.schema),
+        autocompleteQuery: (nextProps.schema.reduce((totalLength, table) =>
+          totalLength + table.columns.length, 0) <= 5000),
+      };
+    }
+    return null;
+  }
+
+  render() {
+    // eslint-disable-next-line react/prop-types
+    const modKey = this.props.KeyboardShortcuts.modKey;
+    const parameterTooltip = <Tooltip id="parameterTooltip">Add New Parameter (<i>{modKey} + P</i>)</Tooltip>;
+    const formatTooltip = <Tooltip id="formatTooltip">Format Query</Tooltip>;
+    const saveTooltip = <Tooltip id="saveTooltip">{modKey} + S</Tooltip>;
+    const executeTooltip = <Tooltip id="executeTooltip">{modKey} + Enter</Tooltip>;
+
+    return (
+      <section style={{ height: '100%' }}>
+        <div className="container p-15 m-b-10" style={{ height: '100%' }}>
+          <div style={{ height: 'calc(100% - 40px)', marginBottom: '0px' }} className="editor__container">
+            <AceEditor
+              ref={this.refEditor}
+              theme="textmate"
+              mode={this.props.dataSource.syntax || 'sql'}
+              value={this.props.queryText}
+              editorProps={{ $blockScrolling: Infinity }}
+              width="100%"
+              height="100%"
+              setOptions={{
+                behavioursEnabled: true,
+                enableSnippets: true,
+                enableLiveAutocompletion: this.state.autocompleteQuery,
+                autoScrollEditorIntoView: true,
+              }}
+              showPrintMargin={false}
+              wrapEnabled={false}
+              onLoad={this.onLoad}
+              onPaste={this.onPaste}
+              onChange={(queryText) => { this.props.updateQuery(queryText); }}
+            />
+          </div>
+
+          <div className="editor__control">
```

Comment:
It does make sense to move the editor controls from `query.html` into their own component, but why should they be part of the query editor component?

Reply:
I need to review the structure of this page to get a good idea, but maybe EditorControls?

Anyway, it shouldn't be coupled with the editor component. What's beautiful in React that you can nicely compose multiple small narrow scoped components.